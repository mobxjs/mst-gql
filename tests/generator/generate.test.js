/// <reference types="jest"/>

const { scaffold } = require("../../generator/generate")
const escapeStringRegexp = require("escape-string-regexp")

const toRegex = snippet => new RegExp(`\\s+${escapeStringRegexp(snippet)}\\s+`)
const findFile = (output, name) =>
  output.find(o => o.length && o.length > 1 && o[0] === name)
const hasFileContent = (file, snippet) => file[1].match(toRegex(snippet))

test("basic scaffolding to work", () => {
  expect(
    scaffold(
      `
type User {
  id: ID
  name: String!
  avatar: String!
}
type Query {
  me: User
}
`,
      { roots: ["User"] }
    )
  ).toMatchSnapshot()
})

test("interface field type to work", () => {
  const output = scaffold(
    `
interface Owner {
  id: ID!
  name: String!
}      
type User implements Owner {
  id: ID!
  name: String!
  avatar: String!
}
type Organization implements Owner {
  id: ID!
  name: String!
  logo: String!
}
type Repo {
  id: ID!
  owner: Owner
}
type Query {
  repo: Repo
}
`,
    { roots: ["Repo"] }
  )
  expect(output).toMatchSnapshot()

  expect(findFile(output, "OwnerModelSelector")).toBeTruthy()

  const repoModelBase = findFile(output, "RepoModel.base")
  expect(repoModelBase).toBeTruthy()
  expect(
    hasFileContent(
      repoModelBase,
      "owner: types.union(types.undefined, types.null, types.union(types.late(() => UserModel), types.late(() => OrganizationModel))),"
    )
  ).toBeTruthy()
})

test("union field type to work", () => {
  const output = scaffold(
    `
type Movie {
  description: String!
  director: String!
}
type Book {
  description: String!
  author: String!
}
union SearchItem = Movie | Book
type SearchResult {
  inputQuery: String!
  items: [SearchItem]!
}
type Query {
  search(text: String!): SearchResult!
}
`,
    { roots: ["SearchResult"] }
  )
  expect(output).toMatchSnapshot()

  expect(findFile(output, "SearchItemModelSelector")).toBeTruthy()

  const searchResultBase = findFile(output, "SearchResultModel.base")
  expect(searchResultBase).toBeTruthy()
  expect(
    hasFileContent(
      searchResultBase,
      "items: types.union(types.undefined, types.array(types.union(types.null, types.union(types.late(() => MovieModel), types.late(() => BookModel))))),"
    )
  ).toBeTruthy()
})
