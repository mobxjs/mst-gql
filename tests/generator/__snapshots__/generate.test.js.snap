// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic scaffolding to work 1`] = `
Array [
  Array [
    "ModelBase",
    "import { MSTGQLObject } from \\"mst-gql\\"

export const ModelBase = MSTGQLObject
",
    false,
  ],
  Array [
    "UserModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { RootStoreType } from \\"./index\\"


/**
 * UserBase
 * auto generated base class for the model UserModel.
 */
export const UserModelBase = ModelBase
  .named('User')
  .props({
    __typename: types.optional(types.literal(\\"User\\"), \\"User\\"),
    id: types.identifier,
    name: types.union(types.undefined, types.string),
    avatar: types.union(types.undefined, types.string),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class UserModelSelector extends QueryBuilder {
  get id() { return this.__attr(\`id\`) }
  get name() { return this.__attr(\`name\`) }
  get avatar() { return this.__attr(\`avatar\`) }
}
export function selectFromUser() {
  return new UserModelSelector()
}

export const userModelPrimitives = selectFromUser().name.avatar
",
    true,
  ],
  Array [
    "UserModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { UserModelBase } from \\"./UserModel.base\\"

/* The TypeScript type of an instance of UserModel */
export interface UserModelType extends Instance<typeof UserModel.Type> {}

/* A graphql query fragment builders for UserModel */
export { selectFromUser, userModelPrimitives, UserModelSelector } from \\"./UserModel.base\\"

/**
 * UserModel
 */
export const UserModel = UserModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore",
    "import { Instance } from \\"mobx-state-tree\\"
import { RootStoreBase } from \\"./RootStore.base\\"

export interface RootStoreType extends Instance<typeof RootStore.Type> {}

export const RootStore = RootStoreBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */
import { types } from \\"mobx-state-tree\\"
import { MSTGQLStore, configureStoreMixin, QueryOptions } from \\"mst-gql\\"

import { UserModel, UserModelType } from \\"./UserModel\\"
import { userModelPrimitives, UserModelSelector } from \\"./UserModel.base\\"


/**
* Store, managing, among others, all the objects received through graphQL
*/
export const RootStoreBase = MSTGQLStore
  .named(\\"RootStore\\")
  .extend(configureStoreMixin([['User', () => UserModel]], ['User']))
  .props({
    users: types.optional(types.map(types.late(() => UserModel)), {})
  })
  .actions(self => ({
    queryMe(variables?: {  }, resultSelector: string | ((qb: UserModelSelector) => UserModelSelector) = userModelPrimitives.toString(), options: QueryOptions = {}) {
      return self.query<{ me: UserModelType}>(\`query me { me {
        \${typeof resultSelector === \\"function\\" ? resultSelector(new UserModelSelector()).toString() : resultSelector}
      } }\`, variables, options)
    },
  }))
",
    true,
  ],
  Array [
    "reactUtils",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { createStoreContext, createUseQueryHook } from \\"mst-gql\\"
import * as React from \\"react\\"
import { RootStore, RootStoreType } from \\"./RootStore\\"

export const StoreContext = createStoreContext<RootStoreType>(React)

export const useQuery = createUseQueryHook(StoreContext, React)
",
    true,
  ],
  Array [
    "index",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

export * from \\"./UserModel\\"
export * from \\"./RootStore\\"
export * from \\"./reactUtils\\"
",
    true,
  ],
]
`;

exports[`interface field type to work 1`] = `
Array [
  Array [
    "ModelBase",
    "import { MSTGQLObject } from \\"mst-gql\\"

export const ModelBase = MSTGQLObject
",
    false,
  ],
  Array [
    "RepoModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { OrganizationModel } from \\"./OrganizationModel\\"
import { OwnerModelSelector } from \\"./OwnerModelSelector\\"
import { UserModel } from \\"./UserModel\\"
import { RootStoreType } from \\"./index\\"


/**
 * RepoBase
 * auto generated base class for the model RepoModel.
 */
export const RepoModelBase = ModelBase
  .named('Repo')
  .props({
    __typename: types.optional(types.literal(\\"Repo\\"), \\"Repo\\"),
    id: types.identifier,
    owner: types.union(types.undefined, types.null, types.union(types.late(() => UserModel), types.late(() => OrganizationModel))),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class RepoModelSelector extends QueryBuilder {
  get id() { return this.__attr(\`id\`) }
  owner(builder?: string | OwnerModelSelector | ((selector: OwnerModelSelector) => OwnerModelSelector)) { return this.__child(\`owner\`, OwnerModelSelector, builder) }
}
export function selectFromRepo() {
  return new RepoModelSelector()
}

export const repoModelPrimitives = selectFromRepo()
",
    true,
  ],
  Array [
    "RepoModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { RepoModelBase } from \\"./RepoModel.base\\"

/* The TypeScript type of an instance of RepoModel */
export interface RepoModelType extends Instance<typeof RepoModel.Type> {}

/* A graphql query fragment builders for RepoModel */
export { selectFromRepo, repoModelPrimitives, RepoModelSelector } from \\"./RepoModel.base\\"

/**
 * RepoModel
 */
export const RepoModel = RepoModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "OwnerModelSelector",
    "import { QueryBuilder } from \\"mst-gql\\"
import { OrganizationModelSelector } from \\"./OrganizationModel.base\\"
import { UserModelSelector } from \\"./UserModel.base\\"

export class OwnerModelSelector extends QueryBuilder {
  get id() { return this.__attr(\`id\`) }
  get name() { return this.__attr(\`name\`) }
  user(builder?: string | UserModelSelector | ((selector: UserModelSelector) => UserModelSelector)) { return this.__inlineFragment(\`User\`, UserModelSelector, builder) }
  organization(builder?: string | OrganizationModelSelector | ((selector: OrganizationModelSelector) => OrganizationModelSelector)) { return this.__inlineFragment(\`Organization\`, OrganizationModelSelector, builder) }
}
export function selectFromOwner() {
  return new OwnerModelSelector()
}

export const ownerModelPrimitives = selectFromOwner().name",
    true,
  ],
  Array [
    "UserModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { RootStoreType } from \\"./index\\"


/**
 * UserBase
 * auto generated base class for the model UserModel.
 */
export const UserModelBase = ModelBase
  .named('User')
  .props({
    __typename: types.optional(types.literal(\\"User\\"), \\"User\\"),
    id: types.identifier,
    name: types.union(types.undefined, types.string),
    avatar: types.union(types.undefined, types.string),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class UserModelSelector extends QueryBuilder {
  get id() { return this.__attr(\`id\`) }
  get name() { return this.__attr(\`name\`) }
  get avatar() { return this.__attr(\`avatar\`) }
}
export function selectFromUser() {
  return new UserModelSelector()
}

export const userModelPrimitives = selectFromUser().name.avatar
",
    true,
  ],
  Array [
    "UserModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { UserModelBase } from \\"./UserModel.base\\"

/* The TypeScript type of an instance of UserModel */
export interface UserModelType extends Instance<typeof UserModel.Type> {}

/* A graphql query fragment builders for UserModel */
export { selectFromUser, userModelPrimitives, UserModelSelector } from \\"./UserModel.base\\"

/**
 * UserModel
 */
export const UserModel = UserModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "OrganizationModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { RootStoreType } from \\"./index\\"


/**
 * OrganizationBase
 * auto generated base class for the model OrganizationModel.
 */
export const OrganizationModelBase = ModelBase
  .named('Organization')
  .props({
    __typename: types.optional(types.literal(\\"Organization\\"), \\"Organization\\"),
    id: types.identifier,
    name: types.union(types.undefined, types.string),
    logo: types.union(types.undefined, types.string),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class OrganizationModelSelector extends QueryBuilder {
  get id() { return this.__attr(\`id\`) }
  get name() { return this.__attr(\`name\`) }
  get logo() { return this.__attr(\`logo\`) }
}
export function selectFromOrganization() {
  return new OrganizationModelSelector()
}

export const organizationModelPrimitives = selectFromOrganization().name.logo
",
    true,
  ],
  Array [
    "OrganizationModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { OrganizationModelBase } from \\"./OrganizationModel.base\\"

/* The TypeScript type of an instance of OrganizationModel */
export interface OrganizationModelType extends Instance<typeof OrganizationModel.Type> {}

/* A graphql query fragment builders for OrganizationModel */
export { selectFromOrganization, organizationModelPrimitives, OrganizationModelSelector } from \\"./OrganizationModel.base\\"

/**
 * OrganizationModel
 */
export const OrganizationModel = OrganizationModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore",
    "import { Instance } from \\"mobx-state-tree\\"
import { RootStoreBase } from \\"./RootStore.base\\"

export interface RootStoreType extends Instance<typeof RootStore.Type> {}

export const RootStore = RootStoreBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */
import { types } from \\"mobx-state-tree\\"
import { MSTGQLStore, configureStoreMixin, QueryOptions } from \\"mst-gql\\"

import { RepoModel, RepoModelType } from \\"./RepoModel\\"
import { repoModelPrimitives, RepoModelSelector } from \\"./RepoModel.base\\"
import { UserModel, UserModelType } from \\"./UserModel\\"
import { userModelPrimitives, UserModelSelector } from \\"./UserModel.base\\"
import { OrganizationModel, OrganizationModelType } from \\"./OrganizationModel\\"
import { organizationModelPrimitives, OrganizationModelSelector } from \\"./OrganizationModel.base\\"


/**
* Store, managing, among others, all the objects received through graphQL
*/
export const RootStoreBase = MSTGQLStore
  .named(\\"RootStore\\")
  .extend(configureStoreMixin([['Repo', () => RepoModel], ['User', () => UserModel], ['Organization', () => OrganizationModel]], ['Repo']))
  .props({
    repos: types.optional(types.map(types.late(() => RepoModel)), {})
  })
  .actions(self => ({
    queryRepo(variables?: {  }, resultSelector: string | ((qb: RepoModelSelector) => RepoModelSelector) = repoModelPrimitives.toString(), options: QueryOptions = {}) {
      return self.query<{ repo: RepoModelType}>(\`query repo { repo {
        \${typeof resultSelector === \\"function\\" ? resultSelector(new RepoModelSelector()).toString() : resultSelector}
      } }\`, variables, options)
    },
  }))
",
    true,
  ],
  Array [
    "reactUtils",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { createStoreContext, createUseQueryHook } from \\"mst-gql\\"
import * as React from \\"react\\"
import { RootStore, RootStoreType } from \\"./RootStore\\"

export const StoreContext = createStoreContext<RootStoreType>(React)

export const useQuery = createUseQueryHook(StoreContext, React)
",
    true,
  ],
  Array [
    "index",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

export * from \\"./RepoModel\\"
export * from \\"./OwnerModelSelector\\"
export * from \\"./UserModel\\"
export * from \\"./OrganizationModel\\"
export * from \\"./RootStore\\"
export * from \\"./reactUtils\\"
",
    true,
  ],
]
`;

exports[`union field type to work 1`] = `
Array [
  Array [
    "ModelBase",
    "import { MSTGQLObject } from \\"mst-gql\\"

export const ModelBase = MSTGQLObject
",
    false,
  ],
  Array [
    "SearchResultModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { BookModel } from \\"./BookModel\\"
import { MovieModel } from \\"./MovieModel\\"
import { SearchItemModelSelector } from \\"./SearchItemModelSelector\\"
import { RootStoreType } from \\"./index\\"


/**
 * SearchResultBase
 * auto generated base class for the model SearchResultModel.
 */
export const SearchResultModelBase = ModelBase
  .named('SearchResult')
  .props({
    __typename: types.optional(types.literal(\\"SearchResult\\"), \\"SearchResult\\"),
    inputQuery: types.union(types.undefined, types.string),
    items: types.union(types.undefined, types.array(types.union(types.null, types.union(types.late(() => MovieModel), types.late(() => BookModel))))),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class SearchResultModelSelector extends QueryBuilder {
  get inputQuery() { return this.__attr(\`inputQuery\`) }
  items(builder?: string | SearchItemModelSelector | ((selector: SearchItemModelSelector) => SearchItemModelSelector)) { return this.__child(\`items\`, SearchItemModelSelector, builder) }
}
export function selectFromSearchResult() {
  return new SearchResultModelSelector()
}

export const searchResultModelPrimitives = selectFromSearchResult().inputQuery
",
    true,
  ],
  Array [
    "SearchResultModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { SearchResultModelBase } from \\"./SearchResultModel.base\\"

/* The TypeScript type of an instance of SearchResultModel */
export interface SearchResultModelType extends Instance<typeof SearchResultModel.Type> {}

/* A graphql query fragment builders for SearchResultModel */
export { selectFromSearchResult, searchResultModelPrimitives, SearchResultModelSelector } from \\"./SearchResultModel.base\\"

/**
 * SearchResultModel
 */
export const SearchResultModel = SearchResultModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "SearchItemModelSelector",
    "import { QueryBuilder } from \\"mst-gql\\"
import { BookModelSelector, bookModelPrimitives } from \\"./BookModel.base\\"
import { MovieModelSelector, movieModelPrimitives } from \\"./MovieModel.base\\"

export class SearchItemModelSelector extends QueryBuilder {
  movie(builder?: string | MovieModelSelector | ((selector: MovieModelSelector) => MovieModelSelector)) { return this.__inlineFragment(\`Movie\`, MovieModelSelector, builder) }
  book(builder?: string | BookModelSelector | ((selector: BookModelSelector) => BookModelSelector)) { return this.__inlineFragment(\`Book\`, BookModelSelector, builder) }
}
export function selectFromSearchItem() {
  return new SearchItemModelSelector()
}

// provides all primitive fields of union member types combined together
export const searchItemModelPrimitives = selectFromSearchItem().movie(movieModelPrimitives).book(bookModelPrimitives)",
    true,
  ],
  Array [
    "MovieModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { RootStoreType } from \\"./index\\"


/**
 * MovieBase
 * auto generated base class for the model MovieModel.
 */
export const MovieModelBase = ModelBase
  .named('Movie')
  .props({
    __typename: types.optional(types.literal(\\"Movie\\"), \\"Movie\\"),
    description: types.union(types.undefined, types.string),
    director: types.union(types.undefined, types.string),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class MovieModelSelector extends QueryBuilder {
  get description() { return this.__attr(\`description\`) }
  get director() { return this.__attr(\`director\`) }
}
export function selectFromMovie() {
  return new MovieModelSelector()
}

export const movieModelPrimitives = selectFromMovie().description.director
",
    true,
  ],
  Array [
    "MovieModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { MovieModelBase } from \\"./MovieModel.base\\"

/* The TypeScript type of an instance of MovieModel */
export interface MovieModelType extends Instance<typeof MovieModel.Type> {}

/* A graphql query fragment builders for MovieModel */
export { selectFromMovie, movieModelPrimitives, MovieModelSelector } from \\"./MovieModel.base\\"

/**
 * MovieModel
 */
export const MovieModel = MovieModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "BookModel.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { types } from \\"mobx-state-tree\\"
import { QueryBuilder } from \\"mst-gql\\"
import { ModelBase } from \\"./ModelBase\\"
import { RootStoreType } from \\"./index\\"


/**
 * BookBase
 * auto generated base class for the model BookModel.
 */
export const BookModelBase = ModelBase
  .named('Book')
  .props({
    __typename: types.optional(types.literal(\\"Book\\"), \\"Book\\"),
    description: types.union(types.undefined, types.string),
    author: types.union(types.undefined, types.string),
  })
  .views(self => ({
    get store() {
      return self.__getStore<RootStoreType>()
    }
  }))

export class BookModelSelector extends QueryBuilder {
  get description() { return this.__attr(\`description\`) }
  get author() { return this.__attr(\`author\`) }
}
export function selectFromBook() {
  return new BookModelSelector()
}

export const bookModelPrimitives = selectFromBook().description.author
",
    true,
  ],
  Array [
    "BookModel",
    "import { Instance } from \\"mobx-state-tree\\"
import { BookModelBase } from \\"./BookModel.base\\"

/* The TypeScript type of an instance of BookModel */
export interface BookModelType extends Instance<typeof BookModel.Type> {}

/* A graphql query fragment builders for BookModel */
export { selectFromBook, bookModelPrimitives, BookModelSelector } from \\"./BookModel.base\\"

/**
 * BookModel
 */
export const BookModel = BookModelBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore",
    "import { Instance } from \\"mobx-state-tree\\"
import { RootStoreBase } from \\"./RootStore.base\\"

export interface RootStoreType extends Instance<typeof RootStore.Type> {}

export const RootStore = RootStoreBase
  .actions(self => ({
    // This is an auto-generated example action.
    log() {
      console.log(JSON.stringify(self))
    }
  }))
",
    false,
  ],
  Array [
    "RootStore.base",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */
import { types } from \\"mobx-state-tree\\"
import { MSTGQLStore, configureStoreMixin, QueryOptions } from \\"mst-gql\\"

import { SearchResultModel, SearchResultModelType } from \\"./SearchResultModel\\"
import { searchResultModelPrimitives, SearchResultModelSelector } from \\"./SearchResultModel.base\\"
import { MovieModel, MovieModelType } from \\"./MovieModel\\"
import { movieModelPrimitives, MovieModelSelector } from \\"./MovieModel.base\\"
import { BookModel, BookModelType } from \\"./BookModel\\"
import { bookModelPrimitives, BookModelSelector } from \\"./BookModel.base\\"


/**
* Store, managing, among others, all the objects received through graphQL
*/
export const RootStoreBase = MSTGQLStore
  .named(\\"RootStore\\")
  .extend(configureStoreMixin([['SearchResult', () => SearchResultModel], ['Movie', () => MovieModel], ['Book', () => BookModel]], ['SearchResult']))
  .props({
    searchresults: types.optional(types.map(types.late(() => SearchResultModel)), {})
  })
  .actions(self => ({
    querySearch(variables: { text: string }, resultSelector: string | ((qb: SearchResultModelSelector) => SearchResultModelSelector) = searchResultModelPrimitives.toString(), options: QueryOptions = {}) {
      return self.query<{ search: SearchResultModelType}>(\`query search($text: String!) { search(text: $text) {
        \${typeof resultSelector === \\"function\\" ? resultSelector(new SearchResultModelSelector()).toString() : resultSelector}
      } }\`, variables, options)
    },
  }))
",
    true,
  ],
  Array [
    "reactUtils",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

import { createStoreContext, createUseQueryHook } from \\"mst-gql\\"
import * as React from \\"react\\"
import { RootStore, RootStoreType } from \\"./RootStore\\"

export const StoreContext = createStoreContext<RootStoreType>(React)

export const useQuery = createUseQueryHook(StoreContext, React)
",
    true,
  ],
  Array [
    "index",
    "/* This is a mst-gql generated file, don't modify it manually */
/* eslint-disable */
/* tslint:disable */

export * from \\"./SearchResultModel\\"
export * from \\"./SearchItemModelSelector\\"
export * from \\"./MovieModel\\"
export * from \\"./BookModel\\"
export * from \\"./RootStore\\"
export * from \\"./reactUtils\\"
",
    true,
  ],
]
`;
